lib/Support/BackedgeBuilder.cpp

```diff
@@ -28,24 +28,40 @@ void Backedge::setValue(mlir::Value newValue) {
 }
 
 BackedgeBuilder::~BackedgeBuilder() {
+  assert(succeeded(clearOrEmitError()) && "backedges still in use");
+}
+
+LogicalResult BackedgeBuilder::clearOrEmitError() {
+  unsigned numInUse = 0;
   for (Operation *op : edges) {
-    assert(op->use_empty() && "Backedge still in use");
+    if (!op->use_empty()) {
+      op->emitError("backedge of type `")
+          << op->getResult(0).getType() << "`still in use";
+      ++numInUse;
+      continue;
+    }
     if (rewriter)
       rewriter->eraseOp(op);
     else
       op->erase();
   }
+  edges.clear();
+  if (numInUse > 0)
+    mlir::emitRemark(loc, "abandoned ") << numInUse << " backedges";
+  return success(numInUse == 0);
 }
 
-Backedge::operator mlir::Value() { return value; }
+void BackedgeBuilder::abandon() { edges.clear(); }
```

Comment:
This implementation doesn't seem to match the method description in the header, depending on what you mean by "cursor ops".

Reply:
This makes the builder forget about the existing edges. It's basically a way of saying "I need to be able to destroy this BackedgeBuilder regardless of whether there are edges". This allows passes to have graceful `return failure()` style aborts out of a scope that contains a `BackedgeBuilder` that might still have open backedges. Without this the builder is significantly less useful because you can't use it in a user-facing pass, since early-outs would always cause the executable to assert, even if you had a lowering failure and the fact that there are unclosed backedges is a lesser concern.

You wouldn't call `abandon()` during normal operation. Its sole purpose is to be an escape hatch in controlled circumstances.

Reply:
If I'm understanding properly, why not just convert the assertion in the deconstructor to an error? That would certainly make debugging easier.

Reply:
Yeah I also thought about that, but the problem there is that you can't detect something has gone wrong from within the pass, or overall when looking at the exit code of the program. In most cases an unfinished backedge is a logical error in a pass that you have to abort pass execution over, either through and assert or by `signalPassFailure()`. Unfortunately the only thing you can do in the destructor is the assert way.

The current implementation preserves the assert in the destructor, but it also tries to emit the errors. So depending on how the diagnostic engine is set up, you get the executable to abort while still seeing the reported errors.